### ISSUE TITLE
Unison server failed: Uncaught exception: Out of memory

### ISSUE THREAD

**Issue by Rotlet** [bcpierce00/unison#1057]
Hi all.
I'm upgrading from Unison 2.51.5 (ocaml 4.11.2) to Unison 2.53.5 (ocaml 4.14.2).
I built the ocaml packages, then built unison from source. Everything went well. Then I stopped the service, replaced the files and started it again, when I tried to start it I got the following error:
unison[2360270]: Unison 2.53.5 (ocaml 4.14.2): Contacting server…
unison[2360270]: Connected [//server2//data/cache -> //server1//data/cache]
unison[2362467]: Looking for changes
unison[2362467]: Unison server failed: Uncaught exception (please report a bug): Out of memory
unison[2362467]: Raised at Lwt.ignore result in file "/root/unison-2.53.5/src/lwt/lwt.ml", line 121, characters 6-13
unison[2362467]: Called from Stdlib List.iter in file "list.ml", line 110, characters 12-15
unison[2362467]: Called from Lwt.restart in file "/root/unison-2.53.5/src/lwt/Iwt.ml", line 31, characters 2-37
unison[2362467]: Called from Lwt.wakeup in file "/root/unison-2.53.5/src/lwt/lwt.ml" (inlined), line 69, characters 17-46
unison[2362467]: Called from Lwt_unix_impl.restart_threads in file "/root/unison-2.53.5/src/lwt/generic/lwt_unix_impl.ml", line 52, characters 6-23
unison[2362467]: Called from Lwt unix impl.run in file "/root/unison-2.53.5/src/Iwt/generic/Iwt unix impl.ml", line 125, characters 6-40
unison[2362467]: Called from Main.init.(fun) in file "/root/unison-2.53.5/src/main.ml", line 256, characters 6-24
unison[2362467]: Called from Main.catchall in file "/root/unison-2.53.5/src/main.ml", line 161, characters 6-10
unison[2362467]: Error: Lost connection with the server
systemd[1]: unison.service: Main process exited, code=exited, status=3/NOTIMPLEMENTED
systemd[1]: unison.service: Failed with result 'exit-code'.
Why might this happen and what can you try to solve it?

Servers are running RHEL 8.5 (with unison service)/8.4, the build was done directly on each server. Unison service parameters:
[Unit]
Description=Unison sync service
After=network-online.target
[Service]
User=unison
Group=unison
TimeoutStartSec=600
RestartSec=600
ExecStart=/usr/local/bin/unison
IgnoreSIGPIPE=false
KillMode=control-group
Restart=always
PIDFile=/var/run/unison-service.pid
LimitMEMLOCK=infinity
Memory Limit=30G
[Install]
WantedBy=multi-user.target

**Comment by gdt** [comment]
Your bug report left out information about the size (files and bytes) of your replica.

This is basically #378.    It's a blend of
  - the error message/behavior should be nicer
  - unison should use less memory
  - your data is big

This is unlikely to be truly a bug (that isn't a dup of #378).   It could perhaps be viewed as a missing feature "Unison should use temp files or be re-coded to handle arbitrary data in finite memory",  but that's basically not going to happen.   It's possible there is a bug that's causing this instead.

Basically, if you want to make progress, you're going to have to debug it, varying your workload, logging how much memory is being used, being clear on the ulimit for memory that is actually in place, etc.  You can start by 1) making sure the remote is also up to date and 2) taking systemd out of the equation.   I don't want to provide debugging help in issues; the issue tracker is geared towards baked bug reports.  Feel free to ask for help on the users list, or hackers list if you are talking about specific bits of code.

**Comment by tleedjarv** [comment]
It could be a regression, seeing as it started happening on upgrade.

@Rotlet please provide as much information as you can about your env (see @gdt's comment). The very first thing you should do is try without systemd and see if you still get the same error.

**Comment by Rotlet** [comment]
Thanks for your advice. I tried to run the command without using systemctl and the synchronization started. I currently have 13m files on one of the servers and their number is constantly growing, the average file size varies from 900b to 21kb. The second server has more files, but I have not tried to run the synchronization there. Do I understand correctly that in order for the synchronization to be repeated constantly, I need to make a cron job with the repeat parameter?

**Comment by gdt** [comment]
Please read the entire manual, and ask for help on the unison-users list.

**Comment by gdt** [comment]
Reading the history of this issue:
  - feedback timeout
  - running not under systemd resulted in unison working
  - seems to be a dup of #378 anyway
Feel free to comment with debugging data, or to ask for help on the user's list.

**Comment by gdt** [comment]
Happy to reopen with good debugging data.   Right now we don't know if systemd is setting limits too low, or where those limits are coming from.   We also don't have "2.51 had a max process size of X while syncing this but 2.53.5 had Y".
